### PR TITLE
New memory instances

### PR DESCRIPTION
**Pull request by terrylucas**

### CODE REVIEW

File: flutter-studio/src/io/flutter/profiler/FilterLibraryDialog.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
+/*
```
[the remark this answers is not in the document]
Comment (terrylucas) [reply flutter-studio/src/io/flutter/profiler/FilterLibraryDialog.java]: No, its the automatic copyright header when I select create new Java class from IntelliJ.  Made it the Chromium copyright.

File: flutter-studio/src/io/flutter/profiler/FilterLibraryDialog.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    JPanel panel = new JPanel();
+    panel.setLayout(new BoxLayout(panel, BoxLayout.PAGE_AXIS));
+
+    popupDialog = new PopupLibraryFilter(allLibraries);
+
+    panel.add(popupDialog);
+
+    Border buttonBorder = BorderFactory.createLineBorder(Color.LIGHT_GRAY);
+
+    CommonButton check = new CommonButton("Check All");
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/FilterLibraryDialog.java]: Nit:
"Check all" to "Select All"

File: flutter-studio/src/io/flutter/profiler/FilterLibraryDialog.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    check.setBorder(buttonBorder);
+    check.addMouseListener(new MouseAdapter() {
+      @Override
+      public void mouseClicked(MouseEvent e) {
+        popupDialog.checkBoxList.setCheckStateAllItems(true);
+        popupDialog.checkBoxList.updateUI();
+      }
+    });
+
+    CommonButton uncheck = new CommonButton("Uncheck All");
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/FilterLibraryDialog.java]: Uncheck All --> Unselect All

File: flutter-studio/src/io/flutter/profiler/FilterLibraryDialog.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    }
+
+    public void setCheckStateAllItems(boolean isChecked) {
+      int modelSize = this.getModel().getSize();
+      for (int index = 0; index < modelSize; index++) {
+        this.getModel().getElementAt(index).setSelected(isChecked);
+      }
+    }
+
+    public void setCheckStateAllItems(List<String> librariesSelected) {
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/FilterLibraryDialog.java]: nit: is there a reason librariesSelected isn't
Set<String> to make this method more efficient?

File: flutter-studio/src/io/flutter/profiler/FlutterAllMemoryData.java [hunk 5]
```diff
@@ -47,9 +49,13 @@ public ThreadSafeData() { }
         if (data.x >= xRange.getMin()) {
           // TODO(terry): Hack for now to get last item so we don't flicker. Consider using a for/loop instead of iterator.
           if (lastData != null && outData.isEmpty()) {
-            outData.add(lastData);
+            // TODO(terry): Need to copy data because of modification done in LineChart for stacked plotting.
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/FlutterAllMemoryData.java]: Looks like you are copying out the data so perhaps remove the TODO(terry): as that made it look like we weren't copying.
Having the comment is great to explain why the data is copied.

File: flutter-studio/src/io/flutter/profiler/FlutterStudioProfilersView.java [hunk 6]
```diff
@@ -79,10 +96,15 @@
   private CommonButton resetZoom;
   private CommonButton frameSelection;
   private ProfilerAction frameSelectionAction;
+  private CommonButton snapshot;      // Snapshot the VM memory.
+  private CommonButton filterLibrary; // Filter libraries
+  private List<String> selectedLibraries;
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/FlutterStudioProfilersView.java]: make this
Set<String>?

File: flutter-studio/src/io/flutter/profiler/FlutterStudioProfilersView.java [hunk 7]
```diff
@@ -211,6 +233,70 @@ private void initializeStageUi() {
...
+        // Update the classes table if visible.  If not, we'll wait for a snapshot click.
+        view.memorySnapshot.removeAllClassChildren(true);
+      }
+
+      Iterator<String> iterator = selectedLibraries.iterator();
+      while (iterator.hasNext()) {
+        String libraryName = iterator.next();
+        LibraryRef libraryRef = view.allLibraries.get(libraryName);
+        if (libraryRef == null) {
+          if (libraryName == "dart:*") {
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/FlutterStudioProfilersView.java]: nit make "dart:*" a constant as this is some strange wild card.

File: src/io/flutter/inspector/HeapState.java [hunk 8]
```diff
@@ -92,36 +95,40 @@ void addSample(HeapMonitor.HeapSample sample) {
...
+        total += heap.getCapacity();
+        external += heap.getExternal();
       }
     }
 
     rssBytes = vm.getJson().get("_currentRSS").getAsInt();
     heapMaxInBytes = total;
 
-    addSample(new HeapMonitor.HeapSample(current, false));
+    addSample(new HeapMonitor.HeapSample(current, external,false));
```
Comment (jacob314) [inline src/io/flutter/inspector/HeapState.java]: run code format on this file and others to fix text formatting issues.

File: flutter-studio/src/io/flutter/profiler/FilterLibraryDialog.java (hunk 4 above)
Comment (terrylucas) [reply flutter-studio/src/io/flutter/profiler/FilterLibraryDialog.java]: None, I suspect I was using List<String> as I was working on the code, moved to the set for filtering but never changed the initial list used by the dialog to set.  Done.

File: flutter-studio/src/io/flutter/profiler/FlutterAllMemoryData.java (hunk 5 above)
Comment (terrylucas) [reply flutter-studio/src/io/flutter/profiler/FlutterAllMemoryData.java]: Right, its an old comment before I copied the data to workaround a problem with the adt-ui code.  I'll remove the comment until I no longer have to copy the data.

File: flutter-studio/src/io/flutter/profiler/FlutterStudioProfilersView.java (hunk 6 above)
Comment (terrylucas) [reply flutter-studio/src/io/flutter/profiler/FlutterStudioProfilersView.java]: Done from previous comment on fixing.

File: src/io/flutter/inspector/HeapState.java (hunk 8 above)
Comment (terrylucas) [reply src/io/flutter/inspector/HeapState.java]: Formatted all files in profiler directory

File: src/io/flutter/inspector/HeapState.java [hunk 9]
```diff
@@ -54,7 +56,8 @@ public int getMaxSampleSizeMs() {
     return samples.samples;
   }
 
-  public int getMaxHeapInBytes() {
+  // Aollocated heap size.
```
[the remark this answers is not in the document]
Comment (terrylucas) [reply src/io/flutter/inspector/HeapState.java]: Fixed

File: flutter-studio/src/io/flutter/profiler/Memory.java [hunk 10]
```diff
@@ -0,0 +1,417 @@
...
+
+    // Update the new data.
+    _myInstancesTreeModel.reload();
+  }
+}
+
+final class TreeUtil {
+  private static boolean ascending;
+
+  // JDK 1.8.0
```
[the remark this answers is not in the document]
Comment (terrylucas) [reply flutter-studio/src/io/flutter/profiler/Memory.java]: Fixed this and removed a number of other v8/v9 comments and some commented out code I was testing.

**Comment by jacob314** [comment]
lgtm

**Pull request merged**